### PR TITLE
Fixing -public .class/.method without trailing spaces

### PR DESCRIPTION
**Pull request by BrianBohe**

### CODE REVIEW

**Comment by BrianBohe** [comment]
I cannot push directly to your repository

File: ILTransform/ILRewriter.cs [hunk 1]
```diff
@@ -97,6 +97,24 @@ private void RewriteFile(string source)
                 int lineIndex = _testProject.LastMainMethodDefLine;
                 if (lineIndex >= 0)
                 {
+                    if (_settings.CollapseMainSign
```
Comment (markples) [inline ILTransform/ILRewriter.cs]: curious - why is this needed?  If we have the bounds of the definition (first and last), then why can't we just add 'public' and be done?

I'm not necessarily against it, but every transformation has some risk of going wrong.

Comment (markples) [reply ILTransform/ILRewriter.cs]: Should that be "CollapseMainLines"?

File: ILTransform/ILRewriter.cs [hunk 2]
```diff
@@ -97,6 +97,24 @@ private void RewriteFile(string source)
...
+                        && (_testProject.FirstMainMethodDefLine != -1)
+                        && (_testProject.FirstMainMethodDefLine < _testProject.LastMainMethodDefLine))
+                    {
+                        int mainLine = _testProject.FirstMainMethodDefLine;
+                        string collapsedLine = lines[mainLine].TrimEnd();
+                        for(mainLine++; mainLine <= _testProject.LastMainMethodDefLine; mainLine++)
+                        {
+                            collapsedLine += " " + lines[mainLine].Trim();
+                        }
```
Comment (markples) [inline ILTransform/ILRewriter.cs]: (not tested)
```suggestion
                        string collapsedLine = string.Join(' ',
                            lines
                                .Skip(_testProject.FirstMainMethodDefLine)
                                .Take(mainLinesLength)
                                .Select((s, i) => i == 0 ? s.TrimEnd() : s.Trim()));
```

File: ILTransform/ILRewriter.cs [hunk 3]
```diff
@@ -97,6 +97,24 @@ private void RewriteFile(string source)
...
+                        // overwrite first main line and remove the rest
+                        lines[_testProject.FirstMainMethodDefLine] = collapsedLine;
+                        int mainSignLineLength = _testProject.LastMainMethodDefLine - _testProject.FirstMainMethodDefLine;
+                        lines.RemoveRange(_testProject.FirstMainMethodDefLine+1, mainSignLineLength);
```
Comment (markples) [inline ILTransform/ILRewriter.cs]: Is this removing an extra line?

File: ILTransform/TestProject.cs [hunk 4]
```diff
@@ -223,15 +223,21 @@ public static bool MakePublic(bool isILTest, ref string line, bool force)
             if (force)
             {
                 int charIndex = line.SkipWhiteSpace();
+                string preffix = line[..charIndex];
```
Comment (markples) [inline ILTransform/TestProject.cs:226]: See `GetIndent` - you'd need a variant that returns int and string or consider `AddAfterIndent`.  (Fix it to do all whitespace so that other callers can benefit too - I have seen tabs in these files)

Comment (markples) [reply ILTransform/TestProject.cs:226]: (AddAfterIndent does not seem right here)

File: ILTransform/TestProject.cs [hunk 5]
```diff
@@ -223,15 +223,21 @@ public static bool MakePublic(bool isILTest, ref string line, bool force)
...
                     if (charIndex < line.Length)
                     {
                         charIndex++;
                     }
                 }
-                line = string.Concat(line.AsSpan(0, charIndex), "public ", line.AsSpan(charIndex));
+                // Avoid white space at the end if line is ".class" or add white space if something continues
+                // after the new "public" token.
+                string suffix = charIndex == line.Length ? "" : string.Concat(" ", line.AsSpan(charIndex));
+                line = string.Concat(preffix, "public", suffix);
```
Comment (markples) [inline ILTransform/TestProject.cs]: This method seems messy trying to combine the C# and IL logic a bit even before your change..

if (IL)
  skip non whitespace
  insert " public"
else
  don't skip
  insert "public "

(btw, string.Insert(x, value) seems simpler than the Concat of prefix, value, suffix though that Concat pattern seems to exist in a lot of places)

(also, "prefix")

File: ILTransform/ILRewriter.cs (hunk 3 above)
Comment (BrianBohe) [reply ILTransform/ILRewriter.cs]: This always confuses me. So, the substruction would give the number of items as if it is [,). For example, if it is a main sign in two lines, `mainSignLineLength = 1` and we have to remove from `_testProject.FirstMainMethodDefLine+1` (the second line)

File: ILTransform/ILRewriter.cs (hunk 1 above)
Comment (BrianBohe) [reply ILTransform/ILRewriter.cs]: I haven't checked other regexes, but this makes the scenario for them easier to fix and understand. There are not so many cases of main signs on multiple lines ,<10 in JIT/Regression

File: ILTransform/TestProject.cs (hunk 5 above)
Comment (BrianBohe) [reply ILTransform/TestProject.cs]: I like your idea, thanks!

File: ILTransform/TestProject.cs (hunk 4 above)
Comment (BrianBohe) [reply ILTransform/TestProject.cs:226]: Thanks! I removed prefix following another comment, but I will take into account GetIndent for next changes

File: ILTransform/ILRewriter.cs (hunk 2 above)
Comment (BrianBohe) [reply ILTransform/ILRewriter.cs]: Thanks I took that!

File: ILTransform/Program.cs [hunk 6]
```diff
@@ -60,6 +61,10 @@ public static int Main(string[] args)
                         {
                             settings.MakePublic = true;
                         }
+                        else if (arg =="-collapse-main-sig")
```
Comment (markples) [inline ILTransform/Program.cs:64]:
```suggestion
                        else if (arg == "-collapse-main-sig")
```

File: ILTransform/ILRewriter.cs [hunk 7]
```diff
@@ -97,6 +97,21 @@ private void RewriteFile(string source)
...
+                        && (_testProject.FirstMainMethodDefLine != -1)
+                        && (_testProject.FirstMainMethodDefLine < _testProject.LastMainMethodDefLine))
+                    {
+                        // overwrite first main line and remove the rest
+                        int mainLinesLength = _testProject.LastMainMethodDefLine + 1 - _testProject.FirstMainMethodDefLine;
+                        lines[_testProject.FirstMainMethodDefLine] = string.Join(' ',
+                               lines.Skip(_testProject.FirstMainMethodDefLine)
+                               .Take(mainLinesLength)
+                               .Select((s, i) => i == 0 ? s.TrimEnd() : s.Trim()));
+                        lines.RemoveRange(_testProject.FirstMainMethodDefLine+1, mainLinesLength - 1);
```
Comment (markples) [inline ILTransform/ILRewriter.cs:110]:
```suggestion
                        lines.RemoveRange(_testProject.FirstMainMethodDefLine + 1, mainLinesLength - 1);
```

**Review by markples — APPROVED** [review]
